### PR TITLE
Antelope : Permission Operations

### PR DESCRIPTION
**Pull request by zolting**
Added the Permission Operations table for Antelope

Caveats : 

- I had some trouble handling the optional nature of the new_perm_last_updated and old_perm_last_updated fields. Not sure what would be the best way to handle this. They're commented out of the schema and table insertion logic for now
- The Authority field in a PermissionObject isn't implemented yet, as it's a quite complex type

Perm ops select

### CODE REVIEW

**Comment by DenisCarriere** [comment]
@zolting if you want to have a look at those conflicts

Other then that, looks good to merge 👍

File: blocks/antelope/src/perm_ops.rs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+pub fn insert_perm_op(tables: &mut DatabaseChanges, clock: &Clock, transaction: &TransactionTrace, perm_op: &PermOp) {
+    // transaction
+    let tx_hash = &transaction.id;
+
+    // action
+    let action_index = perm_op.action_index;
+
+    let old_perm = perm_op.old_perm.clone().unwrap_or_default();
```
Comment (DenisCarriere) [inline blocks/antelope/src/perm_ops.rs:24]: For optional fields, you can use `.as_ref().expect("old_perm missing");`

Example: https://github.com/pinax-network/subgraph-antelope-transactions/blob/c6c20a0e4839eb69a2d3090a7d7a339af5345178/src/transactions.rs#L20

**Review by DenisCarriere — CHANGES_REQUESTED** [review]

**Comment by DenisCarriere** [comment]
The Authority is definitely useful, however I agree it's a bit more of a complex type

image

Definitely need `Array` support for that one or create more tables to load the `authorities`, it would be useful to have `authority` for all the accounts

## `authorities` TABLE
```sql
    accounts_permissions                 Array(String), // [actor@permission]
    accounts_weights                     Array(UInt32), // [weight]
    keys                                 Array(String), // [PUB_K1_73ci5Qnwfcxs2S7LAMpVRcFr9hc3YcfZ7Nh5pSsCvEFKbQGYhL]
    keys_weight                          Array(UInt32), // [weight]
    threshold                            UInt32
    waits_weight                         Array(UInt32), // [weight]
    waits_wait_sec                       Array(UInt32), // [wait_sec]
```

File: blocks/antelope/src/perm_ops.rs (hunk 1 above)
Comment (zolting) [reply blocks/antelope/src/perm_ops.rs:24]: The sink systematically crashes with expect, as there are many perm_ops with no old_perm

**Comment by zolting** [comment]
Added the late_updated fields. You can see the different behaviors possible with the old_perm and new_perm fields here:

old_perm new_perm behaviors

How should we handle the absence of old_perm? Make the clickhouse fields optional? Right now having 0 as value for `old_perm_id` and `old_perm_parent_id` might be a little confusing

**Comment by DenisCarriere** [comment]
1. Added new tables to handle `authority`:
- accounts
- keys
- waits
2. flattened `perm_ops` to only include latest new permission

> ❗ I do not believe that `new_perm` can be empty, since updating any new permissions would always result in a new permission, I don't believe it can be deleted (but need to confirm that)

## `Authority.*` tables
```sql
CREATE TABLE IF NOT EXISTS accounts
(
    `index`                     UInt32,
    action_index                UInt32,
    actor                       String,
    permission                  LowCardinality(String),
    weight                      UInt32,
);

CREATE TABLE keys
(
    `index`                     UInt32,
    action_index                UInt32,
    public_key                  String,
    weight                      UInt32,
);

CREATE TABLE waits
(
    `index`                     UInt32,
    action_index                UInt32,
    wait_sec                    UInt32,
    weight                      UInt32,
)
```

## Updated Perm Ops

```sql

CREATE TABLE perm_ops
(
    operation                   LowCardinality(String),
    operation_code              UInt8,
    action_index                UInt32,
    id                          UInt64,
    parent_id                   UInt64,
    owner                       String,
    name                        String,
    threshold                   UInt32,
);
```

**Review by DenisCarriere — APPROVED** [review]

File: blocks/antelope/src/perm_ops.rs (hunk 1 above)
Comment (DenisCarriere) [reply blocks/antelope/src/perm_ops.rs]: I removed `old_perm` from the schema structure, we only need to handle new permissions

Old permissions can be looked up by past actions

**Pull request merged**